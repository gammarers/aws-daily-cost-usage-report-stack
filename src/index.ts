import * as crypto from 'crypto';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';
import { CostReporterFunction } from './funcs/cost-reporter-function';

export interface DailyCostUsageReporterProps {
  readonly slackWebhookUrl: string;
  readonly slackPostChannel: string;
}

export class DailyCostUsageReporter extends Construct {
  constructor(scope: Construct, id: string, props: DailyCostUsageReporterProps) {
    super(scope, id);

    // ðŸ‘‡Get current account & region
    // const account = cdk.Stack.of(this).account;
    // const region = cdk.Stack.of(this).region;

    // ðŸ‘‡Create random key
    const randomNameKey = crypto.createHash('shake256', { outputLength: 4 })
      .update(`${cdk.Names.uniqueId(scope)}-${cdk.Names.uniqueId(this)}`)
      .digest('hex');

    // ðŸ‘‡Lambda Exec Role
    const lambdaExecutionRole = new iam.Role(this, 'LambdaExecutionRole', {
      roleName: `cost-report-lambda-exec-${randomNameKey}-role`,
      description: '',
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      ],
      inlinePolicies: {
        ['get-cost-usage']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'ce:GetCostAndUsage',
              ],
              resources: ['*'],
            }),
          ],
        }),
      },
    });

    // ðŸ‘‡Lambda Function
    //const lambdaFunction =
    new CostReporterFunction(this, 'CostReporterFunction', {
      functionName: `cost-report-${randomNameKey}-func`,
      description: 'A function to archive logs s3 bucket from CloudWatch Logs.',
      environment: {
        //BUCKET_NAME: logArchiveBucket.bucketName,
        SLACK_WEBHOOK_URL: props.slackWebhookUrl,
        SLACK_POST_CHANNEL: props.slackPostChannel,
      },
      role: lambdaExecutionRole,
    });

    // ðŸ‘‡Schedule

  }

}