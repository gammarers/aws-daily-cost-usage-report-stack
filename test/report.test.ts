import { App } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { CostGroupType, DailyCostUsageReportStack } from '../src';

describe('CostUsageReports Construct Testing', () => {

  describe('CostUsageReport Construct Normal(only required arguments) Testing', () => {

    const app = new App();

    const stack = new DailyCostUsageReportStack(app, 'DailyCostUsageReportStack', {
      slackToken: 'xoxb-11111111111-XXXXXXXXXXXXX-XXXXXXXXXXXXXXXXXXXXXXXX',
      slackChannel: 'example-channel',
      costGroupType: CostGroupType.SERVICES,
    });

    const template = Template.fromStack(stack);

    describe('Lambda Testing', () => {

      it('Should have lambda execution role', () => {
        template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
          RoleName: Match.stringLikeRegexp('cost-report-lambda-exec-.*-role'),
          AssumeRolePolicyDocument: Match.objectEquals({
            Version: '2012-10-17',
            Statement: Match.arrayWith([
              Match.objectEquals({
                Effect: 'Allow',
                Principal: {
                  Service: 'lambda.amazonaws.com',
                },
                Action: 'sts:AssumeRole',
              }),
            ]),
          }),
          ManagedPolicyArns: Match.arrayWith([
            {
              'Fn::Join': Match.arrayEquals([
                '',
                Match.arrayEquals([
                  'arn:',
                  {
                    Ref: 'AWS::Partition',
                  },
                  ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
                ]),
              ]),
            },
          ]),
          Policies: Match.arrayEquals([
            {
              PolicyName: 'get-cost-usage',
              PolicyDocument: Match.objectEquals({
                Version: '2012-10-17',
                Statement: [
                  Match.objectEquals({
                    Effect: 'Allow',
                    Action: 'ce:GetCostAndUsage',
                    Resource: '*',
                  }),
                ],
              }),
            },
          ]),
        }));
      });

      it('Should have lambda function', () => {
        template.hasResourceProperties('AWS::Lambda::Function', Match.objectEquals({
          FunctionName: Match.stringLikeRegexp('cost-report-.*-func'),
          Handler: 'index.handler',
          Runtime: 'nodejs20.x',
          Timeout: 45,
          Code: {
            S3Bucket: Match.anyValue(),
            S3Key: Match.stringLikeRegexp('.*.zip'),
          },
          Description: Match.anyValue(),
          Environment: {
            Variables: {
              AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
              SLACK_TOKEN: Match.anyValue(),
              SLACK_CHANNEL: Match.anyValue(),
            },
          },
          Role: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('LambdaExecutionRole'),
              'Arn',
            ],
          },
        }));
      });
    });

    // Schedule testing
    it('Should have Schedule', () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('daily-cost-report-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(1 9 * * ? *)',
        Target: Match.objectEquals({
          Arn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('CostReporterFunction'),
              'Arn',
            ],
          },
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('SchedulerExecutionRole'),
              'Arn',
            ],
          },
          Input: Match.stringLikeRegexp('{"Type":"(Accounts|Services)"}'),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
      template.resourceCountIs('AWS::Scheduler::Schedule', 1);
    });

    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('report');
    });
  });

  describe('CostUsageReport Construct Normal(all arguments) Testing', () => {

    const app = new App();

    const stack = new DailyCostUsageReportStack(app, 'DailyCostUsageReportStack', {
      costGroupType: CostGroupType.SERVICES,
      slackToken: 'xoxb-11111111111-XXXXXXXXXXXXX-XXXXXXXXXXXXXXXXXXXXXXXX',
      slackChannel: 'example-channel',
      scheduleTimezone: 'Asia/Tokyo',
    });

    const template = Template.fromStack(stack);

    it('Should have Schedule', () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('daily-cost-report-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'Asia/Tokyo',
        ScheduleExpression: 'cron(1 9 * * ? *)',
        Target: Match.objectEquals({
          Arn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('CostReporterFunction.'),
              'Arn',
            ],
          },
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('SchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: Match.stringLikeRegexp('{"Type":"(Accounts|Services)"}'),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
      template.resourceCountIs('AWS::Scheduler::Schedule', 1);
    });

    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('report-timezone');
    });
  });
});