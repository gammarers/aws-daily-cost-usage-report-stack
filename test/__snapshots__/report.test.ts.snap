// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CostUsageReports Testing Should match snapshot: report 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DailyCostUsageReporterCostReporterFunction100AA295": Object {
      "DependsOn": Array [
        "DailyCostUsageReporterLambdaExecutionRoleE9CFFB32",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "HASH.zip",
        },
        "Description": "A function to archive logs s3 bucket from CloudWatch Logs.",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SLACK_POST_CHANNEL": "example-channel",
            "SLACK_WEBHOOK_URL": "https://hooks.slack.com/services/xxxxxxxxxx",
          },
        },
        "FunctionName": "cost-report-d83e5229-func",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DailyCostUsageReporterLambdaExecutionRoleE9CFFB32",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "DailyCostUsageReporterLambdaExecutionRoleE9CFFB32": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "ce:GetCostAndUsage",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "get-cost-usage",
          },
        ],
        "RoleName": "cost-report-lambda-exec-d83e5229-role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
